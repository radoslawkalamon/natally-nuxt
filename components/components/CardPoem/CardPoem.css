.card-poem {
  border-radius: var(--border-radius);
  box-shadow: 0 var(--border-size) 0 var(--color-shadow);
  box-sizing: border-box;
  color: var(--color-white);
  display: block;
  height: 72px;
  max-width: 400px;
  overflow: hidden;
  position: relative;
  text-decoration: none;
  transition: box-shadow var(--time-animation) ease, opacity var(--time-animation) ease;
  width: 100%;
}

.card-poem:hover {
  opacity: 0.9;
  text-decoration: underline;
}

.card-poem:active {
  box-shadow: 0 0 0 var(--color-shadow);
  opacity: 0.75;
}

.card-poem__cover {
  box-shadow: inset 0 0 0 var(--border-size) var(--color-shadow);
  box-sizing: border-box;
  height: 100%;
  left: 0;
  object-fit: cover;
  position: absolute;
  top: 0;
  width: 100%;
  z-index: 1;
}

.card-poem__content {
  align-items: center;
  background-color: var(--color-off-black-transparent);
  box-sizing: border-box;
  display: flex;
  gap: var(--gap);
  height: 100%;
  justify-content: space-between;
  left: 0;
  padding: 1em;
  position: absolute;
  top: 0;
  width: 100%;
  z-index: 2;
}

.card-poem__title {
  margin: 0 !important;
}

.card-poem:hover .card-poem__title {
  text-decoration: underline;
}

.card-poem__meta {
  min-width: 100px;
  text-align: right;
}
