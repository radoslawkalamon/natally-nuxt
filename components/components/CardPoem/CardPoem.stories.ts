import type { Meta, Story } from '@storybook/vue'
import ComponentsCardPoem from '@/components/components/CardPoem/CardPoem.vue'
import { DTOMetaPostPoem } from '@/utils/dto.meta.post.poem'

const meta: Meta = {
  title: 'Components / Card Poem',
  component: ComponentsCardPoem
}
export default meta

const Template: Story = (_args, { argTypes }) => ({
  components: {
    ComponentsCardPoem
  },
  props: Object.keys(argTypes),
  template: '<ComponentsCardPoem v-bind="$props" />'
})

export const Default: Story = Template.bind({})
Default.args = {
  metaPostPoem: new DTOMetaPostPoem({
    createdAt: '2021-11-04T00:00:00+00:00',
    description: 'Test description',
    imageCover: 'poezja-314.test-story.webp',
    imageOpenGraph: 'poezja-314.test-story.opengraph.png',
    path: '/test-url',
    title: 'Test story',
    updatedAt: '2021-11-04T00:00:00+00:00'
  })
}
