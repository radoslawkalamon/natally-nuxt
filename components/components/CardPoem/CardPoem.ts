import Vue from 'vue'
import type { VueConstructor } from 'vue'
import { DTOMetaPostPoem } from '@/utils/dto.meta.post.poem'
import { formatterDateMeta } from '@/utils/formatter.date.meta'
import mixinCommonIntersectionObserverShownOnce from '@/utils/mixin.common.intersectionObserver.shownOnce'

export default (Vue as VueConstructor<
  Vue
  & InstanceType<typeof mixinCommonIntersectionObserverShownOnce>
>).extend({
  name: 'ComponentsCardPoem',
  components: {
    ComponentsCover: () => import(/* webpackChunkName: "components" */'@/components/components/Cover/Cover.vue'),
    ComponentsTextMeta: () => import(/* webpackChunkName: "components" */'@/components/components/TextMeta/TextMeta.vue'),
    ComponentsTitle: () => import(/* webpackChunkName: "components" */'@/components/components/Title/Title.vue')
  },
  mixins: [mixinCommonIntersectionObserverShownOnce],
  props: {
    metaPostPoem: {
      type: DTOMetaPostPoem,
      required: true
    }
  },
  computed: {
    'common/intersectionObserver/cacheKey' (): string {
      return this.metaPostPoem.imageCover
    },
    path (): string {
      return this.metaPostPoem.path
    },
    imagesCover (): string[] {
      return this['common/intersectionObserver/shownOnce']
        ? [this.metaPostPoem.imageCover]
        : []
    },
    title (): string {
      return this.metaPostPoem.title
    },
    datePublished (): string {
      return formatterDateMeta(this.metaPostPoem.createdAt)
    },
    datePublishedLabel (): string {
      return `Opublikowano ${this.datePublished}`
    }
  }
})
