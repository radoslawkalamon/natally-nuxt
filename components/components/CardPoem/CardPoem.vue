<template>
  <NuxtLink
    class="card-poem"
    data-test="components-card-poem"
    :to="url"
  >
    <ComponentsCoverPoem
      class="card-poem__cover"
      :image="coverImage"
    />
    <div class="card-poem__content">
      <ComponentsTitle
        class="card-poem__title"
        :title="title"
        :type="3"
        :shall-show-underscore="false"
      />
      <ComponentsTextMeta
        class="card-poem__meta"
        :text="datePublished"
      />
    </div>
  </NuxtLink>
</template>

<script src="./CardPoem.ts" lang="ts"></script>
<style src="./CardPoem.css" scoped></style>
