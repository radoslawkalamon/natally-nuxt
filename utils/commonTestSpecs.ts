import Vue from 'vue'
import { mount } from '@vue/test-utils'
import { merge } from 'lodash'

export function shallRender (component: typeof Vue, dataTestID: string, defaultOptions?: object): void {
  test('shall render', (): void => {
    const wrapper = mount(component, defaultOptions)
    const isWrapperVisible = wrapper.find(`[data-test="${dataTestID}"]`).isVisible()
    expect(isWrapperVisible).toBeTruthy()
  })
}

export function shallHaveSlot (component: typeof Vue, slot: string, defaultOptions?: object): void {
  test(`shall have slot '${slot}'`, (): void => {
    const text = `This is slot '${slot}' test.`
    const wrapperOptions = merge(defaultOptions, {
      slots: {
        [slot]: text
      }
    })
    const wrapper = mount(component, wrapperOptions)
    expect(wrapper.html()).toContain(text)
  })
}

export function shallHaveStringProp (component: typeof Vue, propName: string, defaultOptions?: object): void {
  test(`shall have string prop '${propName}'`, (): void => {
    const text = `This is prop ${propName} test.`
    const wrapperOptions = merge(defaultOptions, {
      propsData: {
        [propName]: text
      }
    })
    const wrapper = mount(component, wrapperOptions)
    expect(wrapper.html()).toContain(text)
  })
}

export function shallHavePropInfluenceOverClassNames (
  component: typeof Vue,
  propName: string,
  propValue: unknown,
  classNameSuffix: string,
  defaultOptions?: object
) {
  test(`shall have ${typeof propValue} prop '${propName}' influence over class names`, () => {
    const wrapperOptions = merge(defaultOptions, {
      propsData: {
        [propName]: propValue
      }
    })
    const wrapper = mount(component, wrapperOptions)
    const isWrapperHaveClass = wrapper.classes().some(cls => cls.includes(classNameSuffix))
    expect(isWrapperHaveClass).toBeTruthy()
  })
}
