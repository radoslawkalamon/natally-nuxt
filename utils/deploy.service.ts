/* eslint-disable no-console */
import { Client, FileInfo, FTPResponse } from 'basic-ftp'
import { ProgressInfo } from 'basic-ftp/dist/ProgressTracker'
import capitalize from './formatters/capitalize'
import spacelize from './formatters/spacelize'

type DeployServiceConfig = {
  host: string;
  port: number;
  username: string;
  password: string;
  remoteCatalog: string;
  remoteIgnoreFile: string[];
  localCatalog: string;
}

type RemoveStrategies = {
  [key in FileInfo['type']]: (name: FileInfo['name']) => Promise<void | FTPResponse>
}

function decoratorLogCall () {
  return function (_target: any, propertyKey: string, descriptor: PropertyDescriptor) {
    const targetMethod = descriptor.value

    descriptor.value = function (...args: any[]) {
      const formattedName = spacelize(capitalize(propertyKey))
      console.log(`::: ${formattedName} :::`)
      return targetMethod.apply(this, args)
    }

    return descriptor
  }
}

class DeployService {
  readonly host: string;
  readonly port: number;
  readonly username: string;
  readonly password: string;
  readonly remoteCatalog: string;
  readonly remoteIgnoreFile: string[];
  readonly localCatalog: string;
  FTPClient = new Client()

  constructor (config: DeployServiceConfig) {
    this.host = config.host
    this.port = Number(config.port)
    this.username = config.username
    this.password = config.password
    this.remoteCatalog = config.remoteCatalog
    this.remoteIgnoreFile = config.remoteIgnoreFile
    this.localCatalog = config.localCatalog
  }

  public async start (): Promise<void> {
    try {
      await this.openConnection()
      await this.ensureRemoteCatalog()
      await this.openRemoteCatalog()
      await this.cleanRemoteCatalog()
      await this.uploadToRemoteCatalog()
    } catch (e) {
      console.error(e)
      throw e
    } finally {
      this.closeConnection()
    }
  }

  @decoratorLogCall()
  private async openConnection (): Promise<void> {
    await this.FTPClient.access({
      host: this.host,
      port: this.port,
      user: this.username,
      password: this.password,
      secure: false
    })
  }

  @decoratorLogCall()
  private async ensureRemoteCatalog (): Promise<void> {
    await this.FTPClient.ensureDir(this.remoteCatalog)
  }

  @decoratorLogCall()
  private async openRemoteCatalog (): Promise<void> {
    await this.FTPClient.cd(this.remoteCatalog)
  }

  @decoratorLogCall()
  private async cleanRemoteCatalog (): Promise<void> {
    const remoteFiles = await this.FTPClient.list()

    for (const file of remoteFiles) {
      if (this.remoteIgnoreFile.includes(file.name)) {
        continue
      }

      const removeStrategies: RemoveStrategies = {
        0: async () => await Promise.resolve(),
        1: async name => await this.FTPClient.remove(name),
        2: async name => await this.FTPClient.removeDir(name),
        3: async () => await Promise.resolve()
      }
      await removeStrategies[file.type](file.name)
    }
  }

  @decoratorLogCall()
  private async uploadToRemoteCatalog (): Promise<void> {
    this.FTPClient.trackProgress(this.handlerUploadProgress)
    await this.FTPClient.uploadFromDir(this.localCatalog)
    this.FTPClient.trackProgress()
  }

  private handlerUploadProgress (info: ProgressInfo): void {
    if (info.bytes === 0) {
      process.stdout.write(`Uploading file ${info.name}\n`)
    }
  }

  @decoratorLogCall()
  private closeConnection (): void {
    this.FTPClient.close()
  }
}

export default DeployService
