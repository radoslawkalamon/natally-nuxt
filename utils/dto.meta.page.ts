export type DTOMetaPageConstructor = {
  createdAt: string;
  description: string;
  imageOpenGraph?: string;
  path: string;
  title: string;
  updatedAt: string;
}

const getPath = (path: string) => {
  const pathWithoutIndex = path.split('/').filter(i => i !== 'index').join('/')
  return `${pathWithoutIndex}${pathWithoutIndex.at(-1) !== '/' ? '/' : ''}`
}

export class DTOMetaPage {
  readonly createdAt: Date;
  readonly description: string;
  readonly imageOpenGraph: string;
  readonly path: string;
  readonly title: string;
  readonly updatedAt: Date;

  constructor ({
    createdAt,
    description,
    imageOpenGraph = 'common.opengraph.png',
    title,
    updatedAt,
    path
  }: DTOMetaPageConstructor) {
    this.createdAt = new Date(createdAt)
    this.description = description
    this.imageOpenGraph = `/images/${imageOpenGraph}`
    this.path = getPath(path)
    this.title = title
    this.updatedAt = new Date(updatedAt)
  }
}
