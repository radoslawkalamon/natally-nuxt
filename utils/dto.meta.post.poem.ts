import { DTOMetaPage, DTOMetaPageConstructor } from '@/utils/dto.meta.page'

export type DTOMetaPostPoemConstructor = DTOMetaPageConstructor & {
  imageCover: string
  imageOpenGraph: string
}

export class DTOMetaPostPoem extends DTOMetaPage {
  readonly imageCover: string;
  readonly imageOpenGraph: string;

  constructor ({
    createdAt,
    description,
    imageCover,
    imageOpenGraph,
    path,
    title,
    updatedAt
  }: DTOMetaPostPoemConstructor) {
    super({
      createdAt,
      description,
      path,
      title,
      updatedAt
    })

    this.imageCover = `${this.imagePath}${imageCover}`
    this.imageOpenGraph = `${this.imagePath}${imageOpenGraph}`
  }

  get imagePath (): string {
    return '/images/poezja-314/'
  }
}
