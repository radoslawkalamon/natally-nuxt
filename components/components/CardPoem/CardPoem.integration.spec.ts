import CardPoem from '@/components/components/CardPoem/CardPoem.vue'
import { shallPassIntegrationSanityTest } from '@/devtools/jest.common.spec'
import { createDefaultOptionsFactory } from '@/devtools/jest.common.spec.utils'
import { JestMockIntersectionObserver } from '@/devtools/jest.mock.IntersectionObserver'
import { DTOMetaPostPoem } from '@/utils/dto.meta.post.poem'

JestMockIntersectionObserver()

const defaultOptionsFactory = createDefaultOptionsFactory({
  propsData: {
    metaPostPoem: new DTOMetaPostPoem({
      createdAt: '2021-11-01T00:00:00+00:00',
      description: 'Test description',
      imageCover: 'image-cover.webp',
      imageOpenGraph: 'image-open-graph.webp',
      path: '/test-url',
      title: 'Test title',
      updatedAt: '2021-11-01T00:00:00+00:00'
    })
  }
})

describe('Components / Card Poem', () => {
  describe('Integration', () => {
    shallPassIntegrationSanityTest({
      component: CardPoem,
      description: 'not in the viewport',
      options: defaultOptionsFactory()
    })

    shallPassIntegrationSanityTest({
      component: CardPoem,
      description: 'in the viewport',
      options: defaultOptionsFactory({
        data () {
          return {
            'common/intersectionObserver/shownOnce': true
          }
        }
      })
    })
  })
})
