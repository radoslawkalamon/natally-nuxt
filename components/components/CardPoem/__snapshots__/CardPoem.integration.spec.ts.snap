// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components / Card Poem Integration shall pass integration sanity test | in the viewport 1`] = `
"<article class=\\"card-poem\\">
  <nuxtlink-stub to=\\"/test-url/\\" class=\\"card-poem__link\\">
    <div aria-hidden=\\"true\\" class=\\"cover card-poem__cover\\"></div>
    <div class=\\"card-poem__content\\">
      <header class=\\"card-poem__header\\">
        <h1 class=\\"card-poem__title title title--design-3\\">Test title</h1>
      </header>
      <footer class=\\"card-poem__footer\\">
        <p class=\\"text-meta card-poem__meta\\" aria-label=\\"Opublikowano 01.11.2021\\" title=\\"Opublikowano 01.11.2021\\">01.11.2021</p>
      </footer>
    </div>
  </nuxtlink-stub>
</article>"
`;

exports[`Components / Card Poem Integration shall pass integration sanity test | not in the viewport 1`] = `
"<article class=\\"card-poem\\">
  <nuxtlink-stub to=\\"/test-url/\\" class=\\"card-poem__link\\">
    <div aria-hidden=\\"true\\" class=\\"cover card-poem__cover\\"></div>
    <div class=\\"card-poem__content\\">
      <header class=\\"card-poem__header\\">
        <h1 class=\\"card-poem__title title title--design-3\\">Test title</h1>
      </header>
      <footer class=\\"card-poem__footer\\">
        <p class=\\"text-meta card-poem__meta\\" aria-label=\\"Opublikowano 01.11.2021\\" title=\\"Opublikowano 01.11.2021\\">01.11.2021</p>
      </footer>
    </div>
  </nuxtlink-stub>
</article>"
`;
